### PR TITLE
Implement on-demand building for streaming JSON

### PR DESCRIPTION
**Pull request by SupunS**
Fix https://github.com/ballerina-platform/ballerina-lang/issues/9961

With this change, iterating/accessing a JSON streaming from a DB will build only on demand.
eg: After reading the datatable from db and converting it to json,
accessing the 100th element, or iterating up to 100th element from a JSON array containing 262,000 records, will only build (load to memory) the first 100 records.

**Existing Impl:**

old

**New Impl:**

new

### CODE REVIEW

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BStreamingJSON.java [hunk 1]
```diff
@@ -68,11 +80,18 @@ public void serialize(OutputStream outputStream) {
...
-            if (this.values != null) {
-                gen.serialize(this);
-            } else {
-                this.datasource.serialize(gen);
+            gen.writeStartArray();
+
+            // First serialize the values loaded to memory
+            for (int i = 0; i < size; i++) {
+                gen.serialize(values[i]);
+            }
```
Comment (Manuri) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BStreamingJSON.java]: Could have used a foreach loop here.

Comment (SupunS) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BStreamingJSON.java]: We actually cannot. `values` array is an auto growing array, and is always larger than the actual filled-size of the array. Hence using foreach will print the un-filled part of the array as well.

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/JSONDataSource.java [hunk 2]
```diff
@@ -37,4 +38,9 @@
      */
     void serialize(JsonGenerator gen) throws IOException;
 
+    boolean hasNext();
```
Comment (hasithaa) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/model/JSONDataSource.java]: Shall we add documentation this, because it is a public API.

Comment (SupunS) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/model/JSONDataSource.java]: Updated

**Pull request merged**